### PR TITLE
Make @nix json structured build log parsing warn instead of fail

### PR DESCRIPTION
**Pull request by roberth**
# Motivation

A build may log with `@nix something` without the intent to write Nix structured logs, resulting in an obscure error.
This makes it easy to understand and continues the build.

Example:
```bash
echo '@nix ]['
```
```
bad JSON log message from builder: [json.exception.parse_error.101] parse error at line 1, column 1: syntax error while parsing value - unexpected ']'; expected '[', '{', or a literal
@nix ][
```

# Context

Cherry-picked from https://gerrit.lix.systems/c/lix/+/2057
Author @lheckemann 

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: src/libutil/logging.cc [hunk 1]
```diff
@@ -295,37 +295,45 @@ bool handleJSONLogMessage(nlohmann::json & json,
...
+        else if (action == "msg") {
+            std::string msg = json["msg"];
+            logger->log((Verbosity) json["level"], msg);
+        }
 
-    return true;
+        return true;
+    } catch (nlohmann::json::exception &e) {
+        warn(
+            "warning: Unable to handle a JSON message from the builder: %s",
```
Comment (Ericson2314) [inline src/libutil/logging.cc:332]: Don't like how this makes assumptions about where this function is being used.

**Review by Ericson2314 — COMMENTED** [review]
+1 on the idea, just don't like the anti-modular way the warning is done.

File: src/libutil/logging.cc (hunk 1 above)
Comment (roberth) [reply src/libutil/logging.cc]: Spot on. That also fixes a technically incorrect message, although it was probably never seen.
Now:

> ```
>             if (handleJSONLogMessage(s, worker.act, worker.hook->activities, "the build hook", true))
> ```

**Comment by roberth** [comment]
Interestingly, the `post-build-hook` does not have `@nix` parsing logic, so that's what that's not in the diff.

**Comment by Mic92** [comment]
@roberth CI fails.

**Comment by pwaller** [comment]
Forgive me if this is off topic, but this might be of interest here: right now if I run `nix develop -f . hello` with nix 2.24.10 on nixos/nixpkgs@staging I get:
```
error: [json.exception.parse_error.101] parse error at line 38, column 445: syntax error while parsing value - invalid string: control character U+001E (RS) must be escaped to \u001E; last read: '" \n    local sep=\"$1\";\n    local name=\"$2\";\n    local type oldifs;\n    if type=$(declare -p \"$name\" 2> /dev/null); then\n        local -n nameref=\"$name\";\n        case \"${type#* }\" in \n            -A*)\n                echo \"concatStringsSep(): ERROR: trying to use concatStringsSep on an associative array.\" 1>&2;\n                return 1\n            ;;\n            -a*)\n                local IFS='<U+001E>'
```

**Pull request merged**